### PR TITLE
Improve CSV writer tests

### LINKED ISSUE #6836: Many CSV writer tests are redundant
A large part of CSV writer tests is implemented as a single, heavily parameterized, test. This leads to combinatoric explosion of the number of tests, where many instances do not contribute to coverage.

- Some parameters have redundant values (e.g. "NEWLINE" and "<<<<<<" as line terminators);
- Majority of parameters to not interact so there's no reason to test the Cartesian product of their value sets;
- Some corner cases are not included, and may need separate tests (e.g. using '-' for separator);

### PR DESCRIPTION
**Pull request by vuule**
Closes #6836 

Overall reduces test execution time and improves coverage:

- Replace around 2.5K test cases with multiple tests that only vary related options.
- Correctly verify the output with multicharacter `line_terminator` option (not supported by readers).
- Add a `seed` call before the random generator is used in one of the tests.
- Simplify a few tests by removing irrelevant comparisons.
- Use buffer output instead of file in affected tests (could be applied to many more).

### CODE REVIEW

**Review by galipremsagar — APPROVED** [review]

**Comment by vuule** [comment]
rerun tests

**Comment by vuule** [comment]
rerun tests

**Comment by vuule** [comment]
rerun tests

**Comment by vuule** [comment]
rerun tests

**Comment by vuule** [comment]
rerun tests

File: python/cudf/cudf/tests/test_csv.py [hunk 1]
```diff
@@ -1502,27 +1501,43 @@ def test_csv_writer_datetime_data(tmpdir):
     assert_eq(expect, got)
 
 
-@pytest.mark.parametrize("sep", [",", "|", " ", ";", np.str_(",")])
+@pytest.mark.parametrize("line_terminator", ["\r", "\n", "\t", np.str_("\n")])
+@pytest.mark.parametrize("sep", [",", "/", np.str_(",")])
+def test_csv_writer_terminator_sep(line_terminator, sep):
+    df = cudf.from_pandas(make_numpy_mixed_dataframe())
```
Comment (brandon-b-miller) [inline python/cudf/cudf/tests/test_csv.py:1507]: It looks like the data that we're using that comes out of `make_numpy_mixed_dataframe()` is used for multiple tests, you might consider factoring it out and just generating it once, maybe as a pytest fixture

File: python/cudf/cudf/tests/test_csv.py [hunk 2]
```diff
@@ -1532,51 +1547,38 @@ def test_csv_writer_datetime_data(tmpdir):
...
+    cudf_buffer = BytesIO()
+    df.to_csv(cudf_buffer, columns=columns, header=header, index=index)
+    pd_buffer = BytesIO()
+    pdf.to_csv(pd_buffer, columns=columns, header=header, index=index)
+
+    expected = cudf.read_csv(pd_buffer, header=0 if header else None)
+    got = cudf.read_csv(cudf_buffer, header=0 if header else None)
+
+    expected_column_cnt = (
+        (1 if index else 0) + len(columns) if columns else got.shape[1]
```
Comment (brandon-b-miller) [inline python/cudf/cudf/tests/test_csv.py:1563]: Should `expected_column_cnt` be based off `got.shape[1]`, or should it be `expected.shape[1]`?

Comment (vuule) [reply python/cudf/cudf/tests/test_csv.py]: Fixed to use the input dataframe size, thanks for pointing this out!

**Review by brandon-b-miller — APPROVED** [review]

File: python/cudf/cudf/tests/test_csv.py (hunk 1 above)
Comment (vuule) [reply python/cudf/cudf/tests/test_csv.py]: Added fixtures for this. Can you please check if I did this correctly? I don't know much about fixtures.

Comment (brandon-b-miller) [reply python/cudf/cudf/tests/test_csv.py]: LGTM

**Comment by vuule** [comment]
rerun tests

**Comment by kkraus14** [comment]
@gpucibot merge

**Pull request merged**